cmd/util/ledger/migrations/staged_contracts_migration.go

```diff
@@ -312,6 +312,10 @@ func (m *StagedContractsMigration) checkContractUpdateValidity(
 	newCode []byte,
 	oldCode ledger.Value,
 ) error {
+	// Parsing and checking of programs has to be done synchronously.
```

Comment:
how so?

Reply:
this just needs to be "thread-safe", because the runtime environment is shared. runtime environment is not designed to be thread safe (e.g: import resolving, etc)

Reply:
Ah, I see, thanks for the explanation.